File: inst/htmlwidgets/finalchart.js
HTMLWidgets.widget({

  name: 'finalchart',

  type: 'output',

  factory: function(el, width, height) {

    // TODO: define shared variables for this instance

      var chart = null ;

  return {

      renderValue: function(x) {
//
//
      if (chart === null ) {
      var aLittleBit = Math.PI / 100000;
      var now = x.year,
          height = x.height - 50 ,
          width = x.width + 50,
          threshold = x.threshold,
          labelThreshold = x.labelThreshold,
          animationDuration = x.animationDuration,
          margin = x.margin,
          arcPadding = x.arcPadding,
          infoPopupDelay = x.infoPopupDelay,
          arcWidth = x.arcWidth,
          sourcePadding = x.sourcePadding,
          targetPadding = x.targetPadding,
          labelPadding = x.labelPadding,
          curveFactor = x.curveFactor,
          maxRegionsOpen = x.maxRegionsOpen ;


     chart = Globalmigration.chart(x.data, {

        now: now,
        height : height ,
        width : width ,
        animationDuration: animationDuration,
        margin: margin,
        arcPadding: arcPadding,
        arcWidth : arcWidth,
        curveFactor : curveFactor,
        maxRegionsOpen : maxRegionsOpen,
        infoPopupDelay : infoPopupDelay,
        sourcePadding: sourcePadding,
        labelPadding: labelPadding,
        targetPadding: targetPadding,
        element:  el ,


        layout: {
          threshold: threshold,
          labelThreshold: labelThreshold,
          colors: x.data.reg_col

        }
      });

    /* var timeline =  Globalmigration.timeline(chart, {
          now: now,
          element: el
        });
*/
      chart.draw(now) ;

     el.chart = chart ;

      }

    el.chart.draw(x.year) ;

//
//
        // TODO: code to render the widget, e.g.

      },

      resize: function(width, height) {

        // TODO: code to re-render the widget with a new size

      }

    };
  }
});

File: R/finalchart.R
#' Migration Flow Visualisation
#'
#' years :
#'
#' @import htmlwidgets
#'
#' @export
finalchart <- function(message,
                       year = 1995,
                       threshold = 5000,
                       labelThreshold = 5000,
                       animationDuration = 500,
                       maxRegionsOpen = 2,
                       margin = 25,
                       arcPadding = 0.04,
                       infoPopupDelay = 300,
                       arcWidth = 24,
                       width = 800,
                       height = 800,
                       labelPadding = 10,
                       sourcePadding = 3,
                       targetPadding = 20,
                       curveFactor = 0.4,
                       elementId = NULL
                       ) {

  # forward options using x
  x = list ( data = message,
             year = year,
             height = height,
             width = width ,
             arcWidth = arcWidth,
             threshold = threshold,
             animationDuration = animationDuration,
             labelThreshold = labelThreshold,
             maxRegionsOpen = maxRegionsOpen,
             margin = margin,
             infoPopupDelay = infoPopupDelay,
             curveFactor = curveFactor,
             labelPadding = labelPadding,
             sourcePadding = sourcePadding,
             targetPadding = targetPadding,
             arcPadding = arcPadding)

  # create widget
  htmlwidgets::createWidget(
    name = 'finalchart',
    x,
  #  htmlwidgets::sizingPolicy(padding = 0,
   #                           browser.fill = TRUE),
    width = width,
    height = height,
    package = 'finalchart'
  )
}

#' Shiny bindings for finalchart
#'
#' Output and render functions for using finalchart within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param expr An expression that generates a finalchart
#' @param env The environment in which to evaluate \code{expr}.
#' @param quoted Is \code{expr} a quoted expression (with \code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#'
#' @name finalchart-shiny
#'
#' @export
finalchartOutput <- function(outputId, width = '100%', height = '400px'){
  htmlwidgets::shinyWidgetOutput(outputId,'finalchart', width, height, package = 'finalchart')
}

#' @rdname finalchart-shiny
#' @export
renderFinalchart <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, finalchartOutput, env, quoted = TRUE)
}
